src/microbe_stage/editor/TimelineTab.cs

```diff
@@ -0,0 +1,249 @@
+
+            cachedLocalTimelineElements.Add(section);
+            eventsContainer.AddChild(section);
+        }
+
+        ApplyEventsFilter();
+    }
+
+    public void TimelineAutoScrollToCurrentTimePeriod()
+    {
+        var scrollRect = scrollContainer.GetGlobalRect();
+        var anchorRect = new Rect2(Vector2.Zero, Vector2.Zero);
+
+        if (eventFilter == Filters.Global)
+        {
+            var last = cachedGlobalTimelineElements.LastOrDefault();
+            anchorRect = last != null ? last.HeaderGlobalRect : new Rect2(Vector2.Zero, Vector2.Zero);
+        }
+        else if (eventFilter == Filters.Local)
+        {
+            var last = cachedLocalTimelineElements.LastOrDefault();
+            anchorRect = last != null ? last.HeaderGlobalRect : new Rect2(Vector2.Zero, Vector2.Zero);
+        }
+
+        var diff = Mathf.Max(Mathf.Min(anchorRect.Position.y, scrollRect.Position.y), anchorRect.Position.y +
+            anchorRect.Size.y - scrollRect.Size.y + (scrollRect.Size.y - anchorRect.Size.y));
+
+        scrollContainer.ScrollVertical += (int)(diff - scrollRect.Position.y);
+    }
+
+    private void ApplyEventsFilter()
+    {
+        switch (EventFilter)
+        {
+            case Filters.Global:
+                cachedGlobalTimelineElements?.ForEach(e => e.Show());
+                cachedLocalTimelineElements?.ForEach(e => e.Hide());
+                globalFilterButton.Pressed = true;
+                break;
+            case Filters.Local:
+                cachedGlobalTimelineElements?.ForEach(e => e.Hide());
+                cachedLocalTimelineElements?.ForEach(e => e.Show());
+                localFilterButton.Pressed = true;
+                break;
+            default:
+                throw new Exception("Not a valid event filter");
+        }
+
+        Invoke.Instance.Queue(TimelineAutoScrollToCurrentTimePeriod);
+    }
+
+    private void OnFilterSelected(int index)
+    {
+        if (!Enum.IsDefined(typeof(Filters), index) || (Filters)index == EventFilter)
+            return;
+
+        GUICommon.Instance.PlayButtonPressSound();
+
+        EventFilter = (Filters)index;
+    }
+
+    private class TimelineSection : VBoxContainer
+    {
+        private Control headerContainer;
+
+        private PackedScene customRtlScene;
+        private StyleBoxTexture eventHighlightStylebox;
+
+        private (double TimePeriod, List<GameEventDescription> Events) data;
+
+        public TimelineSection((double TimePeriod, List<GameEventDescription> Events) data)
+        {
+            this.data = data;
+        }
+
+        public Rect2 HeaderGlobalRect => headerContainer.GetGlobalRect();
+
+        public override void _Ready()
+        {
+            customRtlScene = GD.Load<PackedScene>("res://src/gui_common/CustomRichTextLabel.tscn");
+            eventHighlightStylebox = GD.Load<StyleBoxTexture>(
+                "res://src/microbe_stage/editor/TimelineEventHighlight.tres");
```

Comment:
I think it'd be slightly nicer if these were loaded by the Timeline object so that we don't need to load these 100 times (even though Godot probably caches these, I'd feel better not loading these a ton).